File: apps/app/src/hooks/market-vision/divergence-engine.ts
'use client'

// Generic tolerance-pairing divergence engine.
//
// Extracted verbatim from rsi-divergences.ts so the same algorithm can run
// against any oscillator (RSI, WaveTrend wt2, Stoch K, ...). Unlike the
// VuManChu fractal detector (vmc-divergences.ts), which requires the price
// and oscillator pivots to land on the exact same bar, this engine finds
// price pivots and oscillator pivots independently and pairs them when they
// occur within `tolBars` of each other ("TV-like" mode) — catching the many
// real divergences where the two pivots are 1-2 bars apart.

export type DivergenceType = 'bullish' | 'bearish' | 'h_bullish' | 'h_bearish'

export interface PairedDivergence {
  type: DivergenceType
  startIndex: number
  endIndex: number
  oscStart: number
  oscEnd: number
  priceStart: number
  priceEnd: number
}

export interface DivergenceEngineConfig {
  leftBars: number
  rightBars: number
  pairMode: 'TV-like' | 'Same Bar'
  tolBars: number
  allowEqual: boolean
  priceEps: number
  oscEps: number
  showRegular: boolean
  showHidden: boolean
}

export const DEFAULT_DIVERGENCE_ENGINE_CONFIG: DivergenceEngineConfig = {
  leftBars: 5,
  rightBars: 5,
  pairMode: 'TV-like',
  tolBars: 2,
  allowEqual: true,
  priceEps: 0,
  oscEps: 0,
  showRegular: true,
  showHidden: true,
}

function greater(a: number, b: number, eps: number, allowEqual: boolean): boolean {
  return allowEqual ? a >= b - eps : a > b + eps
}

function less(a: number, b: number, eps: number, allowEqual: boolean): boolean {
  return allowEqual ? a <= b + eps : a < b - eps
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value)
}

export function pivotHighAt(values: number[], leftBars: number, rightBars: number, currentIndex: number): number | null {
  const pivotIndex = currentIndex - rightBars
  if (pivotIndex - leftBars < 0) return null
  if (pivotIndex + rightBars >= values.length) return null

  const pivotValue = values[pivotIndex]
  if (!isFiniteNumber(pivotValue)) return null

  for (let i = pivotIndex - leftBars; i <= pivotIndex + rightBars; i++) {
    if (i === pivotIndex) continue
    const v = values[i]
    if (!isFiniteNumber(v)) return null
    if (v >= pivotValue) return null
  }

  return pivotValue
}

export function pivotLowAt(values: number[], leftBars: number, rightBars: number, currentIndex: number): number | null {
  const pivotIndex = currentIndex - rightBars
  if (pivotIndex - leftBars < 0) return null
  if (pivotIndex + rightBars >= values.length) return null

  const pivotValue = values[pivotIndex]
  if (!isFiniteNumber(pivotValue)) return null

  for (let i = pivotIndex - leftBars; i <= pivotIndex + rightBars; i++) {
    if (i === pivotIndex) continue
    const v = values[i]
    if (!isFiniteNumber(v)) return null
    if (v <= pivotValue) return null
  }

  return pivotValue
}

function expirePending(idxNow: number | null, idxPend: number | null, tolBars: number): boolean {
  if (idxNow == null || idxPend == null) return false
  return idxNow - idxPend > tolBars
}

/**
 * Find regular and hidden divergences between a price series (highs/lows)
 * and an oscillator, pairing independently-detected pivots within a bar
 * tolerance. Bar-index based; callers map indices to times.
 */
export function findPairedDivergences(
  highs: number[],
  lows: number[],
  oscValues: number[],
  config: DivergenceEngineConfig,
): PairedDivergence[] {
  const divergences: PairedDivergence[] = []
  const n = Math.min(highs.length, lows.length, oscValues.length)

  // Pending pairing state (mirrors Pine variables)
  let pendRhiIx: number | null = null
  let pendRhi: number | null = null
  let pendPhiIx: number | null = null
  let pendPhi: number | null = null

  let pendRloIx: number | null = null
  let pendRlo: number | null = null
  let pendPloIx: number | null = null
  let pendPlo: number | null = null

  // Previous paired pivots (for divergence comparisons)
  let prevHighPH: number | null = null
  let prevHighRH: number | null = null
  let prevHighIx: number | null = null

  let prevLowPL: number | null = null
  let prevLowRL: number | null = null
  let prevLowIx: number | null = null

  for (let i = 0; i < n; i++) {
    const ph = pivotHighAt(highs, config.leftBars, config.rightBars, i)
    const pl = pivotLowAt(lows, config.leftBars, config.rightBars, i)
    const rh = pivotHighAt(oscValues, config.leftBars, config.rightBars, i)
    const rl = pivotLowAt(oscValues, config.leftBars, config.rightBars, i)

    const phix = ph == null ? null : i - config.rightBars
    const plix = pl == null ? null : i - config.rightBars
    const rhix = rh == null ? null : i - config.rightBars
    const rlix = rl == null ? null : i - config.rightBars

    // Expire old pending pivots
    if (rhix != null && expirePending(rhix, pendPhiIx, config.tolBars)) {
      pendPhiIx = null
      pendPhi = null
    }
    if (phix != null && expirePending(phix, pendRhiIx, config.tolBars)) {
      pendRhiIx = null
      pendRhi = null
    }
    if (rlix != null && expirePending(rlix, pendPloIx, config.tolBars)) {
      pendPloIx = null
      pendPlo = null
    }
    if (plix != null && expirePending(plix, pendRloIx, config.tolBars)) {
      pendRloIx = null
      pendRlo = null
    }

    // High-side pairing
    let pairHighNow = false
    let curHighPH: number | null = null
    let curHighRH: number | null = null
    let curHighIx: number | null = null

    const phNow = ph != null && phix != null
    const rhNow = rh != null && rhix != null

    if (config.pairMode === 'Same Bar') {
      if (phNow && rhNow && phix === rhix) {
        pairHighNow = true
        curHighPH = ph
        curHighRH = rh
        curHighIx = rhix
      }
    } else {
      if (phNow && rhNow && Math.abs(phix - rhix) <= config.tolBars) {
        pairHighNow = true
        curHighPH = ph
        curHighRH = rh
        curHighIx = rhix
      } else {
        if (rhNow) {
          if (pendPhiIx != null && pendPhi != null && Math.abs(pendPhiIx - rhix) <= config.tolBars) {
            pairHighNow = true
            curHighPH = pendPhi
            curHighRH = rh
            curHighIx = rhix
            pendPhiIx = null
            pendPhi = null
          } else {
            pendRhiIx = rhix
            pendRhi = rh
          }
        }
        if (phNow && !pairHighNow) {
          if (pendRhiIx != null && pendRhi != null && Math.abs(pendRhiIx - phix) <= config.tolBars) {
            pairHighNow = true
            curHighPH = ph
            curHighRH = pendRhi
            curHighIx = pendRhiIx
            pendRhiIx = null
            pendRhi = null
          } else {
            pendPhiIx = phix
            pendPhi = ph
          }
        }
      }
    }

    // Low-side pairing
    let pairLowNow = false
    let curLowPL: number | null = null
    let curLowRL: number | null = null
    let curLowIx: number | null = null

    const plNow = pl != null && plix != null
    const rlNow = rl != null && rlix != null

    if (config.pairMode === 'Same Bar') {
      if (plNow && rlNow && plix === rlix) {
        pairLowNow = true
        curLowPL = pl
        curLowRL = rl
        curLowIx = rlix
      }
    } else {
      if (plNow && rlNow && Math.abs(plix - rlix) <= config.tolBars) {
        pairLowNow = true
        curLowPL = pl
        curLowRL = rl
        curLowIx = rlix
      } else {
        if (rlNow) {
          if (pendPloIx != null && pendPlo != null && Math.abs(pendPloIx - rlix) <= config.tolBars) {
            pairLowNow = true
            curLowPL = pendPlo
            curLowRL = rl
            curLowIx = rlix
            pendPloIx = null
            pendPlo = null
          } else {
            pendRloIx = rlix
            pendRlo = rl
          }
        }
        if (plNow && !pairLowNow) {
          if (pendRloIx != null && pendRlo != null && Math.abs(pendRloIx - plix) <= config.tolBars) {
            pairLowNow = true
            curLowPL = pl
            curLowRL = pendRlo
            curLowIx = pendRloIx
            pendRloIx = null
            pendRlo = null
          } else {
            pendPloIx = plix
            pendPlo = pl
          }
        }
      }
    }

    // Divergence checks — explicit narrowing for type-safety.
    if (
      pairHighNow &&
      curHighIx != null &&
      curHighPH != null &&
      curHighRH != null &&
      prevHighIx != null &&
      prevHighPH != null &&
      prevHighRH != null
    ) {
      const isBearReg =
        config.showRegular &&
        greater(curHighPH, prevHighPH, config.priceEps, config.allowEqual) &&
        less(curHighRH, prevHighRH, config.oscEps, config.allowEqual)

      const isBearHid =
        config.showHidden &&
        less(curHighPH, prevHighPH, config.priceEps, config.allowEqual) &&
        greater(curHighRH, prevHighRH, config.oscEps, config.allowEqual)

      if (isBearReg || isBearHid) {
        divergences.push({
          type: isBearReg ? 'bearish' : 'h_bearish',
          startIndex: prevHighIx,
          endIndex: curHighIx,
          oscStart: prevHighRH,
          oscEnd: curHighRH,
          priceStart: prevHighPH,
          priceEnd: curHighPH,
        })
      }
    }

    if (
      pairLowNow &&
      curLowIx != null &&
      curLowPL != null &&
      curLowRL != null &&
      prevLowIx != null &&
      prevLowPL != null &&
      prevLowRL != null
    ) {
      const isBullReg =
        config.showRegular &&
        less(curLowPL, prevLowPL, config.priceEps, config.allowEqual) &&
        greater(curLowRL, prevLowRL, config.oscEps, config.allowEqual)

      const isBullHid =
        config.showHidden &&
        greater(curLowPL, prevLowPL, config.priceEps, config.allowEqual) &&
        less(curLowRL, prevLowRL, config.oscEps, config.allowEqual)

      if (isBullReg || isBullHid) {
        divergences.push({
          type: isBullReg ? 'bullish' : 'h_bullish',
          startIndex: prevLowIx,
          endIndex: curLowIx,
          oscStart: prevLowRL,
          oscEnd: curLowRL,
          priceStart: prevLowPL,
          priceEnd: curLowPL,
        })
      }
    }

    // Update prev pairs after checks (matches Pine order)
    if (pairHighNow && curHighIx != null && curHighPH != null && curHighRH != null) {
      prevHighPH = curHighPH
      prevHighRH = curHighRH
      prevHighIx = curHighIx
    }
    if (pairLowNow && curLowIx != null && curLowPL != null && curLowRL != null) {
      prevLowPL = curLowPL
      prevLowRL = curLowRL
      prevLowIx = curLowIx
    }
  }

  return divergences
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Indicators/DivergenceEngine.swift
import Foundation

/// Port of `divergence-engine.ts`: tolerance-pairing divergence finder.
public enum DivergenceType: String, Sendable, Codable, Hashable { case bullish, bearish, h_bullish, h_bearish
  public var isBullish: Bool { self == .bullish || self == .h_bullish }
}

public struct PairedDivergence: Sendable, Hashable {
  public var type: DivergenceType
  public var startIndex: Int
  public var endIndex: Int
  public var oscStart: Double
  public var oscEnd: Double
  public var priceStart: Double
  public var priceEnd: Double
}

public enum PairMode: String, Sendable, Codable { case tvLike = "TV-like", sameBar = "Same Bar" }

public struct DivergenceEngineConfig: Sendable, Hashable {
  public var leftBars: Int
  public var rightBars: Int
  public var pairMode: PairMode
  public var tolBars: Int
  public var allowEqual: Bool
  public var priceEps: Double
  public var oscEps: Double
  public var showRegular: Bool
  public var showHidden: Bool
  public init(leftBars: Int, rightBars: Int, pairMode: PairMode, tolBars: Int, allowEqual: Bool, priceEps: Double, oscEps: Double, showRegular: Bool, showHidden: Bool) {
    self.leftBars = leftBars; self.rightBars = rightBars; self.pairMode = pairMode; self.tolBars = tolBars; self.allowEqual = allowEqual
    self.priceEps = priceEps; self.oscEps = oscEps; self.showRegular = showRegular; self.showHidden = showHidden
  }
}

public enum DivergenceEngine {
  static func greater(_ a: Double, _ b: Double, _ eps: Double, _ allowEqual: Bool) -> Bool { allowEqual ? a >= b - eps : a > b + eps }
  static func less(_ a: Double, _ b: Double, _ eps: Double, _ allowEqual: Bool) -> Bool { allowEqual ? a <= b + eps : a < b - eps }

  public static func pivotHighAt(_ values: [Double], leftBars: Int, rightBars: Int, currentIndex: Int) -> Double? {
    let p = currentIndex - rightBars
    guard p - leftBars >= 0, p + rightBars < values.count else { return nil }
    let pv = values[p]
    guard pv.isFinite else { return nil }
    for i in (p - leftBars)...(p + rightBars) where i != p {
      let v = values[i]
      if !v.isFinite || v >= pv { return nil }
    }
    return pv
  }

  public static func pivotLowAt(_ values: [Double], leftBars: Int, rightBars: Int, currentIndex: Int) -> Double? {
    let p = currentIndex - rightBars
    guard p - leftBars >= 0, p + rightBars < values.count else { return nil }
    let pv = values[p]
    guard pv.isFinite else { return nil }
    for i in (p - leftBars)...(p + rightBars) where i != p {
      let v = values[i]
      if !v.isFinite || v <= pv { return nil }
    }
    return pv
  }

  static func expirePending(_ now: Int?, _ pend: Int?, _ tol: Int) -> Bool {
    guard let now, let pend else { return false }
    return now - pend > tol
  }

  public static func findPairedDivergences(highs: [Double], lows: [Double], osc: [Double], config c: DivergenceEngineConfig) -> [PairedDivergence] {
    var out: [PairedDivergence] = []
    let n = min(highs.count, lows.count, osc.count)
    var pendRhiIx: Int? = nil, pendRhi: Double? = nil, pendPhiIx: Int? = nil, pendPhi: Double? = nil
    var pendRloIx: Int? = nil, pendRlo: Double? = nil, pendPloIx: Int? = nil, pendPlo: Double? = nil
    var prevHighPH: Double? = nil, prevHighRH: Double? = nil, prevHighIx: Int? = nil
    var prevLowPL: Double? = nil, prevLowRL: Double? = nil, prevLowIx: Int? = nil

    for i in 0..<n {
      let ph = pivotHighAt(highs, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i)
      let pl = pivotLowAt(lows, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i)
      let rh = pivotHighAt(osc, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i)
      let rl = pivotLowAt(osc, leftBars: c.leftBars, rightBars: c.rightBars, currentIndex: i)
      let phix: Int? = ph == nil ? nil : i - c.rightBars
      let plix: Int? = pl == nil ? nil : i - c.rightBars
      let rhix: Int? = rh == nil ? nil : i - c.rightBars
      let rlix: Int? = rl == nil ? nil : i - c.rightBars

      if rhix != nil, expirePending(rhix, pendPhiIx, c.tolBars) { pendPhiIx = nil; pendPhi = nil }
      if phix != nil, expirePending(phix, pendRhiIx, c.tolBars) { pendRhiIx = nil; pendRhi = nil }
      if rlix != nil, expirePending(rlix, pendPloIx, c.tolBars) { pendPloIx = nil; pendPlo = nil }
      if plix != nil, expirePending(plix, pendRloIx, c.tolBars) { pendRloIx = nil; pendRlo = nil }

      // High side
      var pairHighNow = false
      var curHighPH: Double? = nil, curHighRH: Double? = nil, curHighIx: Int? = nil
      let phNow = ph != nil, rhNow = rh != nil
      if c.pairMode == .sameBar {
        if phNow, rhNow, phix == rhix { pairHighNow = true; curHighPH = ph; curHighRH = rh; curHighIx = rhix }
      } else {
        if phNow, rhNow, abs(phix! - rhix!) <= c.tolBars {
          pairHighNow = true; curHighPH = ph; curHighRH = rh; curHighIx = rhix
        } else {
          if rhNow {
            if let pIx = pendPhiIx, let pv = pendPhi, abs(pIx - rhix!) <= c.tolBars {
              pairHighNow = true; curHighPH = pv; curHighRH = rh; curHighIx = rhix; pendPhiIx = nil; pendPhi = nil
            } else { pendRhiIx = rhix; pendRhi = rh }
          }
          if phNow, !pairHighNow {
            if let rIx = pendRhiIx, let rv = pendRhi, abs(rIx - phix!) <= c.tolBars {
              pairHighNow = true; curHighPH = ph; curHighRH = rv; curHighIx = rIx; pendRhiIx = nil; pendRhi = nil
            } else { pendPhiIx = phix; pendPhi = ph }
          }
        }
      }

      // Low side
      var pairLowNow = false
      var curLowPL: Double? = nil, curLowRL: Double? = nil, curLowIx: Int? = nil
      let plNow = pl != nil, rlNow = rl != nil
      if c.pairMode == .sameBar {
        if plNow, rlNow, plix == rlix { pairLowNow = true; curLowPL = pl; curLowRL = rl; curLowIx = rlix }
      } else {
        if plNow, rlNow, abs(plix! - rlix!) <= c.tolBars {
          pairLowNow = true; curLowPL = pl; curLowRL = rl; curLowIx = rlix
        } else {
          if rlNow {
            if let pIx = pendPloIx, let pv = pendPlo, abs(pIx - rlix!) <= c.tolBars {
              pairLowNow = true; curLowPL = pv; curLowRL = rl; curLowIx = rlix; pendPloIx = nil; pendPlo = nil
            } else { pendRloIx = rlix; pendRlo = rl }
          }
          if plNow, !pairLowNow {
            if let rIx = pendRloIx, let rv = pendRlo, abs(rIx - plix!) <= c.tolBars {
              pairLowNow = true; curLowPL = pl; curLowRL = rv; curLowIx = rIx; pendRloIx = nil; pendRlo = nil
            } else { pendPloIx = plix; pendPlo = pl }
          }
        }
      }

      if pairHighNow, let cIx = curHighIx, let cPH = curHighPH, let cRH = curHighRH, let pIx = prevHighIx, let pPH = prevHighPH, let pRH = prevHighRH {
        let bearReg = c.showRegular && greater(cPH, pPH, c.priceEps, c.allowEqual) && less(cRH, pRH, c.oscEps, c.allowEqual)
        let bearHid = c.showHidden && less(cPH, pPH, c.priceEps, c.allowEqual) && greater(cRH, pRH, c.oscEps, c.allowEqual)
        if bearReg || bearHid {
          out.append(PairedDivergence(type: bearReg ? .bearish : .h_bearish, startIndex: pIx, endIndex: cIx, oscStart: pRH, oscEnd: cRH, priceStart: pPH, priceEnd: cPH))
        }
      }
      if pairLowNow, let cIx = curLowIx, let cPL = curLowPL, let cRL = curLowRL, let pIx = prevLowIx, let pPL = prevLowPL, let pRL = prevLowRL {
        let bullReg = c.showRegular && less(cPL, pPL, c.priceEps, c.allowEqual) && greater(cRL, pRL, c.oscEps, c.allowEqual)
        let bullHid = c.showHidden && greater(cPL, pPL, c.priceEps, c.allowEqual) && less(cRL, pRL, c.oscEps, c.allowEqual)
        if bullReg || bullHid {
          out.append(PairedDivergence(type: bullReg ? .bullish : .h_bullish, startIndex: pIx, endIndex: cIx, oscStart: pRL, oscEnd: cRL, priceStart: pPL, priceEnd: cPL))
        }
      }
      if pairHighNow, let cIx = curHighIx, let cPH = curHighPH, let cRH = curHighRH { prevHighPH = cPH; prevHighRH = cRH; prevHighIx = cIx }
      if pairLowNow, let cIx = curLowIx, let cPL = curLowPL, let cRL = curLowRL { prevLowPL = cPL; prevLowRL = cRL; prevLowIx = cIx }
    }
    return out
  }
}
